content/Django tips and good practices - Bruno Michetti.md

```diff
@@ -0,0 +1,164 @@
+
+## Suggested scaffolding and configuration structure
+As I mentioned, you can generate quickly a base project with the named frameworks. In the documentation you can take a look at the commands to do it. A good practice is to have a structure that increases maintainability, and allow you to easily get a different configuration for each environment (development, production, testing, etc) you use. By default in Django you have a *settings.py* file with all the configuration of the project, and that implies several *if's* separating each part for each environment. A better way to do it is this: You can put the general configuration (common to all environments) in a base file, and then you can create different files corresponding to the different environments. Let's see a simple example, where I changed the file *settings.py* in my project named **example_project**, for a *settings* folder containing the next files:
+
++ \_\_init\_\_.py
++ base.py
++ development.py
++ production.py
+
+The \_\_init\_\_.py file is the one who defines wich configuration use, depending on the environment:
+
+```
+import os
+import importlib
+
+# import specific settings depending on the environment
+ENV_ROLE = os.getenv('ENV_ROLE', 'development')
+
+env_settings = importlib.import_module(f'example_project.settings.{ENV_ROLE}')
+
+globals().update(vars(env_settings))
+
+try:
+    # import local settings if present
+    from .local import *  # noqa
+except ImportError:
+    pass
+```
+So, with this file, you can define on each environment, an environment variable named **ENV_ROLE**, where the value will be the name of the configuration file corresponding to that environment (if you don't define it, the `development` value will be used). After that, you can set in the base.py file the common configuration to all environments (Django INSTALLED_APPS, MIDDLEWARE, TEMPLATES location, etc). And finally, you can separate the configuration of the different environment, for example different databases in development and production, or different allowed hosts, etc.
+
+ 
+For the other hand, in a Django project you can have several applications. You can see an app in Django as a set of related functionalities and models. Another good practice is to have a scaffolding that allows you to see where the apps are, and define standard locations for the functions used by several apps, etc. 
+In short, an orderly and easy to maintain structure. This is a suggested scaffolding:
+
+```
+project_name
+|
+|__project_name (Here you should put all the settings for the app)
+|    |__ __init__.py
+|    |__settings
+|          |__ __init__.py
+|          |__ base.py
+|          |__ development.py
+|          |__ staging.py
+|          |__ production.py
+|          ...
+|    |__wsgi.py(optional)
+|    |__asgi.py(optional)
+|    |__urls.py
+|    ...
+|
+|__templates (Here are general templates, will be loaded last in the application)
```

Comment:
I think i'ts important to explain with more depth the template loading order, and how can you alter it.